frontend/src/components/widgets/DataFrame/hooks/useTableSizer.ts

```diff
@@ -0,0 +1,160 @@
+const MIN_TABLE_WIDTH = MIN_COLUMN_WIDTH + 3
+// Min height for the resizable table container:
+// Based on header + one column, and + 2 for borders + 1 to prevent overlap problem with selection ring.
+const MIN_TABLE_HEIGHT = 2 * ROW_HEIGHT + 3
+const DEFAULT_TABLE_HEIGHT = 400
+
+export type AutoSizerReturn = {
+  rowHeight: number
+  minHeight: number
+  maxHeight: number
+  minWidth: number
+  maxWidth: number
+  resizableSize: ResizableSize
+  setResizableSize: React.Dispatch<React.SetStateAction<ResizableSize>>
+}
+
+export function calculateMaxHeight(numRows: number): number {
+  // +2 pixels for borders
+  return Math.max(numRows * ROW_HEIGHT + 1 + 2, MIN_TABLE_HEIGHT)
+}
+/**
+ * A custom React hook that manages all aspects related to the size of the table.
+ *
+ * @param element - The ArrowProto element
+ * @param numRows - The number of rows in the table
+ * @param containerWidth - The width of the surrounding container
+ * @param containerHeight - The height of the surrounding container
+ * @param isFullScreen - Whether the table is in fullscreen mode
+ *
+ * @returns The row height, min/max height & width, and the current size of the resizable container.
+ */
+function useTableSizer(
+  element: ArrowProto,
+  numRows: number,
+  containerWidth: number,
+  containerHeight?: number,
+  isFullScreen?: boolean
+): AutoSizerReturn {
+  let maxHeight = calculateMaxHeight(
+    numRows +
+      1 + // Column header row
+      (element.editingMode === ArrowProto.EditingMode.DYNAMIC ? 1 : 0) // Trailing row
+  )
+
+  let initialHeight = Math.min(maxHeight, DEFAULT_TABLE_HEIGHT)
+
+  if (element.height) {
+    // User has explicitly configured a height
+    initialHeight = Math.max(element.height, MIN_TABLE_HEIGHT)
+    maxHeight = Math.max(element.height, maxHeight)
+  }
+
+  if (containerHeight) {
+    // If container height is set (e.g. when used in fullscreen)
+    // The maxHeight and height should not be larger than container height
+    initialHeight = Math.min(initialHeight, containerHeight)
+    maxHeight = Math.min(maxHeight, containerHeight)
+
+    if (!element.height) {
+      // If no explicit height is set, set height to max height (fullscreen mode)
+      initialHeight = maxHeight
+    }
+  }
+
+  let initialWidth: number | undefined // If container width is undefined, auto set based on column widths
+  let maxWidth = containerWidth
+
+  if (element.useContainerWidth) {
+    // Always use the full container width
+    initialWidth = containerWidth
+  } else if (element.width) {
+    // User has explicitly configured a width
+    initialWidth = Math.min(
+      Math.max(element.width, MIN_TABLE_WIDTH),
+      containerWidth
+    )
+    maxWidth = Math.min(Math.max(element.width, maxWidth), containerWidth)
+  }
+
+  const [resizableSize, setResizableSize] = React.useState<ResizableSize>({
+    width: initialWidth || "100%",
+    height: initialHeight,
+  })
+
+  React.useLayoutEffect(() => {
```

Comment:
Might be worthwhile to explain the need for `useLayoutEffect` here as a less than common hook. My guess from https://kentcdodds.com/blog/useeffect-vs-uselayouteffect we need this for some DOM work?

Reply:
hmm, good question. So the resizing does have an immediate impact on the DOM, but I'm not sure if it is really needed in this case and I can't remember the exact reason why I choose this. I added that as a fast follow to evaluate if we can change this to `useEffect`. For now, it seems to work stable, so I don't want to touch this before the release.